pylibs/unittests/test_real.py

```diff
@@ -0,0 +1,345 @@
+        self.time_limit = time_limit
+
+    def run(self):
+        start_time = time.time()
+        try:
+            for event in self.stream:
+                if time.time() - start_time > self.time_limit:
+                    self.exception_queue.put(errors.ExpectationError(
+                            "Expectation not fulfilled within time limit"))
+                if (all(s in str(event) for s in self.contain)
+                    and not any(s in str(event) for s in self.not_contain)):
+                    return
+        except grpc._channel._MultiThreadedRendezvous as error:
+            self.exception_queue.put(
+                    errors.ExpectationError("gRPC service terminated"))
+        except grpc._channel._InactiveRpcError as error:
+            self.exception_queue.put(
+                    errors.ExpectationError("gRPC service terminated"))
+
+
+class GRPCClient(object):
+    """Listener for gRPC Visualize stream.
+    """
+    def __init__(self):
+        channel = grpc.insecure_channel("localhost:8999")
+        self.stub = visualize_grpc_pb2_grpc.VisualizeGrpcServiceStub(channel)
+        self.vis_stream = self.stub.Visualize(
+                visualize_grpc_pb2.VisualizeRequest())
+
+    def send_command(self, command: str):
+        self.stub.Command(
+        visualize_grpc_pb2.CommandRequest(command=command))
+
+    def expect_response(self,
+                        contain: List[str]=[],
+                        not_contain: List[str]=[]) -> GRPCThread:
+        exception_queue = queue.Queue()
+        expect_thread = GRPCThread(
+                self.vis_stream, contain, not_contain, exception_queue)
+        expect_thread.start()
+        return exception_queue, expect_thread
+
+
+class BasicTests(OTNSTestCase):
+    def setUp(self) -> None:
+        self.ns = OTNS(otns_args=[
+                "-raw", "-real",
+                "-ot-cli", "otns-silk-proxy",
+                "-listen", ":9000",
+                "-log", "debug"])
+        time.sleep(0.5)
+        self.grpc_client = GRPCClient()
+        self.udp_signalers = {}
+
+    def tearDown(self):
+        for udp_signaler in self.udp_signalers.values():
+            udp_signaler.close()
+
+        super().tearDown()
+
+    def wait_for_expect(self, exception_queue, expect_thread):
+        while expect_thread.is_alive():
+            try:
+                exception = exception_queue.get(block=False)
+            except queue.Empty:
+                pass
+            else:
+                self.fail(exception)
+            
+            expect_thread.join(0.1)
+
+    def testAddNode(self):
+        node_id = random.randint(1, 10)
+        exception_queue, expect_thread = self.grpc_client.expect_response(
+            ["add_node", f"node_id: {node_id}", "x: 100", "y: 100"])
+        self.grpc_client.send_command(f"add router x 100 y 100 id {node_id}")
+        self.goConservative(0.1)
+        self.wait_for_expect(exception_queue, expect_thread)
```

Comment:
Seems such code pattern is used many times. Can we simplify into:
```suggestion
        self.expect_response(f"add router x 100 y 100 id {node_id}", ["add_node", f"node_id: {node_id}", "x: 100", "y: 100"])
```

Reply:
So this middle part includes possible actions like calling the gRPC client and sending UDP messages, therefore the simple refactor you suggested doesn't work across all cases. I experimented with
```Python
def expect_response(self,
                    contain: List[str]=None,
                    not_contain: List[str]=None,
                    action=None,
                    go_step: float=0.1):
    exception_queue, expect_thread = self.grpc_client.expect_response(
        contain=contain,
        not_contain=not_contain)
    action()
    self.goConservative(go_step)
    self.wait_for_expect(exception_queue, expect_thread)
```
And then using it with
```Python
self.expect_response(
        contain=["on_ext_addr_change", f"node_id: {node_id}",
                 f"ext_addr: {extaddr:d}"],
        action=lambda: signaler.send_message(f"extaddr={extaddr:016x}"))
```
But it becomes a lot of hassle for use cases like `testUpdateChildren`, where `action` includes multiple function calls, e.g.
```Python
exception_queue, expect_thread = self.grpc_client.expect_response(
    ["remove_child_table", f"node_id: {node_id_1}",
    f"ext_addr: {extaddr_2:d}"])
signaler_1.send_message("role=3")
signaler_2.send_message("role=1")
signaler_1.send_message(f"child_removed={extaddr_2:016x}")
self.goConservative(0.1)
self.wait_for_expect(exception_queue, expect_thread)
```
In such cases I need to define a separate local function to be passed as `action` argument, and this kind of defeats the purpose to DRY out this code snippet. What do you think?

Reply:
Can we instead define the statuses to be a list ?

Reply:
What about the ones that sends gRPC commands, not statuses?

Reply:
And also status sent from different signalers.

Reply:
I think we can wrap similar simple cases and leave the complex cases expanded.